src/Symfony/Component/Mailer/DelayedSmtpEnvelope.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Mailer;
+
+use Symfony\Component\Mailer\Exception\InvalidArgumentException;
+use Symfony\Component\Mailer\Exception\LogicException;
+use Symfony\Component\Mime\Address;
+use Symfony\Component\Mime\Header\Headers;
+use Symfony\Component\Mime\Message;
+use Symfony\Component\Mime\RawMessage;
+
+/**
+ * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @experimental in 4.3
+ *
+ * @internal
+ */
+final class DelayedSmtpEnvelope extends SmtpEnvelope
+{
+    private $senderSet = false;
+    private $recipientsSet = false;
+    private $message;
+
+    public function __construct(RawMessage $message)
+    {
+        if (!$message instanceof Message) {
```

Comment:
Wouldn't it be better to change the type-hint instead? If we want to support any other message in the future we can still change the type then.

Reply:
I don't think so. I've done it that way to get a "better" error message.

Reply:
But it means that the user will only get an error at runtime if they didn't discover before that `Message` instances are required here. With a proper type hint they can be warned before.

Reply:
How would they be warned before? I suppose you're thinking about static analysis. Not sure what to do.

Reply:
In any case, that's independent from this PR as I'm doing the same at several other places (and this code is copy/pasted from SmtpEnvelope).

Reply:
@xabbuh see #31837